active/0000-addon-instrumentation-experimental-hooks.md

```diff
@@ -0,0 +1,373 @@
+- Start Date: Wednesday, 14 December 2016
+- RFC PR: 
+- Ember CLI Issue: 
+
+# Summary
+
+Add build and load performance hooks that are available to addons.  This enables
+users to write addons that do things like summarize and report build performance
+information.
+
+see https://github.com/ember-cli/ember-cli/issues/6349 for additional context.
+
+# Motivation
+
+Build performance is important to users.  We want to enable users to:
+
+1. Easily discover which portions of their build are costly;
+2. Be able to summarize and report build information in an addon;
+3. Be able to write addons that analyze build performance instrumentation so
+   that they can more easily help diagnose build performance issues in projects
+   to which they do not have direct access.  This is of particular interest to
+   @ember-cli/core &c.
+
+In order to provide these hooks to enable iteration and experimentation prior to
+making firm commitments to format, this rfc propose to initially expose them as
+experiments (see the experiments section below).
+
+# Detailed design
+
+## Experiments
+
+Experiments live in `lib/experiments/index.js`.  Unlike feature flags, there is
+no need to strip them from production.  Experiments allow us to provide power
+user features that are not fully stable without their resorting to private API
+usage.
+
+Experiments are available only in canary builds.  This is achieved by only
+including `lib/experiements/index.js` in canary, and making it the entry point
+for all experiments.
+
+## Build & Load Instrumentation Hooks
+
+We have already a build instrumentation hook as an
+experiment in https://github.com/ember-cli/ember-cli/pull/6546
+
+The goal of this RFC is:
+
+  1. To make the concept of experiments supported
+  2. To promote this particular experiment to public API
+
+### Enabling Instrumentation
+
+Instrumentation is enabled if either the environment variable `BROCCOLI_VIZ` is
+set to `1` or if `EMBER_CLI_INSTRUMENTATION` is set to `1`.
+
+If `BROCCOLI_VIZ=1` then in addition to instrumentation hooks being invoked, a
+serialized form of the instrumentation information is written to disk, that is
+appropriate for consumption by [broccoli-viz](https://github.com/ember-cli/broccoli-viz) which is the current behaviour.
+
+### Build Instrumentation
+
+#### Hook
+
+An addon that implements `buildInstrumentation` will have this hook invoked when
+instrumentation is enabled.
+
+```js
+module.exports = {
+  name: 'my-great-addon',
+
+  buildInstrumentation(buildInstrumentation) {
```

Comment:
In general, we cannot make putting a `buildInstrumentation` method on an addon in older versions of ember-cli throw, but I have implemented a sample that shows how to know that the hook is running ([see here](https://github.com/rwjblue/ember-cli-build-visualizer/blob/309ed77ba78c249085643e26a7b3cf44f6568399/index.js)) by emitting the total number of build steps to the console. This is a totally silly example, just showing that it is possible to use the experiment and properly know it is or isn't functioning (the example prints a `"Nice message here!"`).